Rewrite and centralise generic healthcheck docs

benthorner opened the pull request:
https://trello.com/c/zrq6ih4v/528-investigate-removal-of-no-processes-found-for-x-for-apps-with-a-healthcheck

Please see the commits for more details.

1pretz1 commented on source/manual/alerts/whitehall-app-healthcheck-not-ok.html.md:
```diff
@@ -18,8 +20,8 @@ Production.
-- [Integration][]
-- [Staging][]
+- [Run in Integration Jenkins][]
+- [Run in Staging Jenkins][]
```
These two links are now broken

benthorner replied:
Oops, fixed!

1pretz1 reviewed (commented):
Nice work consolidating these docs @benthorner, much better!

Another thing I thought of concerning the generic app healthcheck doc, could we add a section in around timeouts? They seem to be a fairly common alert and it'd be good to have some information / steps on what to do if a healthcheck is timing out.

Some ideas for the steps could be:

* checking the app is running (upstart, unicorn etc)
* checking the app is serving requests (kibana, app or nginx logs)
* if so checking for high load / traffic and slow requests

benthorner commented on source/manual/alerts/app-healthcheck-not-ok.html.md:
````diff
@@ -0,0 +1,98 @@
+
+```
+# SSH on to a machine running Content Publisher
+gds govuk connect -e integration backend
````
Oops, thanks for spotting. Fixed all these.

benthorner commented on source/manual/alerts/app-healthcheck-not-ok.html.md:
```diff
@@ -0,0 +1,98 @@
+gds govuk connect -e integration backend
+
+# Find the port it's running on
+ps -ef | grep content-publisher
```
👍

benthorner commented on source/manual/alerts/app-healthcheck-not-ok.html.md:
````diff
@@ -0,0 +1,98 @@
+curl localhost:3221/healthcheck
+```
+
+Apps with a custom healthcheck endpoint often make use of [the generic checks in govuk_app_config](https://github.com/alphagov/govuk_app_config/blob/master/docs/healthchecks.md). Some apps also implement custom checks, and the alert should link to custom documentation to explain these.
````
Sure, I've gone with one in Content Publisher.

benthorner commented on source/manual/alerts/app-healthcheck-not-ok.html.md:
````diff
@@ -0,0 +1,98 @@
+- Check for network connectivity to the DB.
+
+```
+# SSH on to machine with problem
````
You no like terse English? Updated.

benthorner commented:
Thanks @1pretz1! I've addressed all the individual comments. Good point about the timeout: I've added a couple of sections at the top for the two kinds of error that you might see.

Could you take another look?

1pretz1 commented on source/manual/alerts/app-healthcheck-not-ok.html.md:
```diff
@@ -0,0 +1,112 @@
+
+## Timeout Error
+
+This means the app is responding to requests, but the request is taking too long (over [20 seconds](https://github.com/alphagov/govuk-puppet/blob/fab936cb82be7fad42636fcafca3718a8368ebfe/modules/icinga/files/usr/lib/nagios/plugins/check_json_healthcheck#L155)).
```
> This means the app is responding to requests

I don't think this is necessarily true. In the recent incident where email-alert-api was failing to start the healthcheck was timing out even though the app wasn't (couldn't) respond to requests. I believe the healthcheck requests made it into Nginx but then timed out waiting to communicate with the app.

Maybe just remove that statement ?

benthorner replied:
That's true. I've amended to say "accepting" requests, but not necessarily responding: the unicorn worker starts and opens a connection, but then encounters a fatal error and exits (hence the timeout). 

I've also added an extra "check the logs" step to cover this.

1pretz1 commented:
Looks good, just one minor comment that needs addressing before 🚢

1pretz1 reviewed (approved):
💯
